core/shared/src/main/scala/zio/ZKeyedPool.scala

```diff
@@ -0,0 +1,181 @@
+                                value = MapValue.Pending(promise)
+                                val previous = map.putIfAbsent(key, value)
+                                if (previous eq null) {
+                                  restore(
+                                    scope
+                                      .extend(
+                                        ZPool
+                                          .make(
+                                            get(key).provideEnvironment(environment),
+                                            range(key),
+                                            ttl(key).getOrElse(Duration.Infinity)
+                                          )
+                                      )
+                                  ).foldCauseZIO(
+                                    cause => {
+                                      map.remove(key, value)
+                                      promise.failCause(cause) *> ZIO.failCause(cause)
+                                    },
+                                    pool => {
+                                      map.put(key, MapValue.Complete(pool))
+                                      promise.succeed(pool).as(pool)
+                                    }
+                                  )
+                                } else {
+                                  previous match {
+                                    case MapValue.Complete(pool)   => ZIO.succeedNow(pool)
+                                    case MapValue.Pending(promise) => restore(promise.await)
+                                  }
+                                }
+                              }
+                            } else {
+                              value match {
+                                case MapValue.Complete(pool)   => ZIO.succeedNow(pool)
+                                case MapValue.Pending(promise) => promise.await
+                              }
+                            }
+                          }
+      activePools = ZIO.suspendSucceed {
+                      ZIO.foreach(Chunk.fromJavaIterable(map.values)) {
+                        case MapValue.Complete(pool)   => ZIO.succeedNow(pool)
+                        case MapValue.Pending(promise) => promise.await
+                      }
+                    }
+    } yield DefaultKeyedPool(getOrCreatePool, activePools)
+
+  private final case class DefaultKeyedPool[Err, Key, Item](
+    getOrCreatePool: Key => ZIO[Any, Nothing, ZPool[Err, Item]],
+    activePools: ZIO[Any, Nothing, Chunk[ZPool[Err, Item]]]
+  ) extends ZKeyedPool[Err, Key, Item] {
+
+    override def get(key: Key)(implicit trace: Trace): ZIO[Scope, Err, Item] =
+      getOrCreatePool(key).flatMap(_.get)
+
+    def invalidate(item: Item)(implicit trace: Trace): UIO[Unit] =
+      activePools.flatMap(ZIO.foreachDiscard(_)(_.invalidate(item)))
```

Comment:
Do you think it does not worth the complexity of tracking which item came from which pool, like in my original version? This adds the item to all pool's invalidated list which affects every pool's next `acquire` I think.

Reply:
I don't think it is sound to do that because the same `Item` could exist in multiple pools. I think the contract of `invalidate` is that it invalidates the `Item` in all pools.

Reply:
that makes sense